### PR TITLE
Add sessionStorage support in widget

### PR DESCRIPTION
**Pull request by sarojkpr**
Fixes:
- Extending support for `sessionStorage` option from widget layer
- Code restructure/refactor to support both `sessionStorage`  and  `localStorage`
- Exposing new event to remove widget data from storage

### CODE REVIEW

**Comment by nareshme2016** [comment]
I think we decided to use local storage only, why are we going back to session storage? how that data will be retained when browser is closed?

**Comment by sarojkpr** [comment]
> I think we decided to use local storage only, why are we going back to session storage? how that data will be retained when browser is closed?

Unless we provide provision for customers to end chat in popout mode, retaining the storage in popout will be a blocker for customers to start a new chat, if they want to.

File: chat-widget/package.json [hunk 1]
```diff
@@ -70,7 +70,8 @@
     "ts-loader": "^9.2.6",
     "typescript": "4.6.4",
     "webpack": "^4.44.2",
-    "webpack-cli": "^4.9.2"
+    "webpack-cli": "^4.9.2",
+     "swiper": "^9.0.5"
```
Comment (charliewang95) [inline chat-widget/package.json:74]: nit: indent issues

File: chat-widget/src/common/storage/default/defaultClientDataStoreProvider.ts [hunk 2]
```diff
@@ -39,7 +45,7 @@ export const defaultClientDataStoreProvider = (cacheTtlinMins = 0): IContextData
                         };
                         const strItem = JSON.stringify(item);
 
-                        if (type === "localStorage") {
+                        if (storageType === StorageType.localStorage) {
                             localStorage.setItem(key, strItem);
                         } else {
                             sessionStorage.setItem(key, strItem);
```
Comment (elopezanaya) [inline chat-widget/src/common/storage/default/defaultClientDataStoreProvider.ts:51]: this block is repeated 3 times, could we consider abstracting it?

**Review by elopezanaya — CHANGES_REQUESTED** [review]

File: chat-widget/src/common/interfaces/IContextDataStore.ts [hunk 3]
```diff
@@ -3,16 +3,16 @@ export interface IContextDataStore {
...
      */
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    setData: (key: string, value: any, type: string) => void;
+    setData: (key: string, value: any) => void;
 
     /**
      * removeData: Remove data from data store by key
      */
-    removeData: (key: string, type: string) => void;
+    removeData: (key: string) => void;
```
[the remark this answers is not in the document]
Comment (sarojkpr) [reply chat-widget/src/common/interfaces/IContextDataStore.ts:17]: we want to keep the localStorage/sessionStorage OOB method name parity

File: chat-widget/src/common/Constants.ts [hunk 4]
```diff
@@ -121,6 +121,8 @@ export class Constants {
     public static readonly PostChatLoadingDurationInMs = 2000;
     public static readonly BrowserUnloadConfirmationMessage = "Do you want to leave chat?";
     public static readonly CacheTtlInMinutes = 15;
+    public static readonly SessionCacheSuffix = "session";
+    public static readonly PopoutCacheSuffix = "popout";
```
[the remark this answers is not in the document]
Comment (sarojkpr) [reply chat-widget/src/common/Constants.ts:125]: Its just an internal constant and we hash it later after being appended to other string, so no way to see this openly.

File: chat-widget/src/common/storage/default/defaultClientDataStoreProvider.ts [hunk 5]
```diff
@@ -6,8 +6,14 @@ import { IContextDataStore } from "../../interfaces/IContextDataStore";
 import { TelemetryHelper } from "../../telemetry/TelemetryHelper";
 import { inMemoryDataStore } from "./defaultInMemoryDataStore";
 
-export const defaultClientDataStoreProvider = (cacheTtlinMins = 0): IContextDataStore => {
+export enum StorageType {
+    "localStorage",
+    "sessionStorage"
+}
```
[the remark this answers is not in the document]
Comment (sarojkpr) [reply chat-widget/src/common/storage/default/defaultClientDataStoreProvider.ts]: fixed.

**Review by elopezanaya — APPROVED** [review]

**Pull request closed without merging**